Add a new scheduling queue based on priority queue.

bsalamat opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
This PR is a part of solution to fix potential starvation of pods in pod preemption. It adds a different scheduling queue to the scheduler that allows highest priority pods to be scheduled before other pods.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
I tried enabling this queue and made sure that our existing tests pass with this queue enabled.

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Add a new scheduling queue that helps schedule the highest priority pending pod first.
```

/sig scheduling

ref/ #54501
ref/ #47604

bsalamat committed:
Added a new scheduling queue based on priority queue.

bsalamat commented:
I enabled the new queue in the code to see if all of our tests pass. The results show that they do pass.

bsalamat commented:
/assign @davidopp 

This is now ready for review.

bsalamat commented:
/retest

timothysc commented on plugin/pkg/scheduler/core/scheduling_queue_test.go:
```diff
@@ -0,0 +1,303 @@
+	}
+}
+
+func TestPriorityQueue_Pop(t *testing.T) {
```
I can't tell what this test does that isn't covered by the previous test.

bsalamat replied:
The unit-test with an underscore in their names test the same function in the code. This one tests PriorityQueue.Pop(). Pop() is sort of tested in other tests, but this one in particular tests that if Pop() is called before Add() (with some probability), pop blocks until the next item is added and returns it.

timothysc commented on plugin/pkg/scheduler/core/scheduling_queue_test.go:
```diff
@@ -0,0 +1,303 @@
+	}
+	if _, exists, _ := q.activeQ.Get(&highPriorityPod); exists {
+		t.Errorf("Didn't expect %v to be in activeQ.", highPriorityPod.Name)
+	}
```
Could also add a pop at the end to verify that only the unschedulablePod exists.

bsalamat replied:
This is testing Delete(). So, I wanted to focus on Delete() only in this test.

timothysc commented on plugin/pkg/scheduler/core/scheduling_queue_test.go:
```diff
@@ -0,0 +1,303 @@
+	}
+}
+
+func TestPriorityQueue_AssignedPodAdded(t *testing.T) {
```
I'm a fan of describing the conditions being tested with comments, as I have to task switch often and it takes a while to unpack some of this.

bsalamat replied:
The test is testing PriorityQueue.AssignedPodAdded(). Are you suggesting to add a test scenario?

bsalamat replied:
I added more comments.

timothysc reviewed (commented):
minor comments

anfernee commented on plugin/pkg/scheduler/core/scheduling_queue.go:
```diff
@@ -15,47 +15,564 @@ limitations under the License.
+
+// Heap is a thread-safe producer/consumer queue that implements a heap data structure.
+// It can be used to implement priority queues and similar data structures.
+type Heap struct {
```
a little confused by this `heapData` and `Heap`. it looks like `heapData` is a pure heap, but `Heap` is a lot more than a heap. maybe it could have another name.

bsalamat replied:
As I mentioned in the comments, this is pretty much the same code that I [implemented in `cache` package](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/heap.go) last quarter. `headData` has the data that `Heap` needs. It also implements golang `heap` interface.
I could have used the same package, but we need to have our own synchronization at `PriorityQueue` level. So, using the package would have done double synchronization and would have imposed extra overhead. That's why I decided to bring the code and modify it as little as possible.
The `Heap` has nothing more than `heapData`. So, I would say `Heap` is still a good name for the struct.

anfernee commented on plugin/pkg/scheduler/core/scheduling_queue.go:
```diff
@@ -15,47 +15,564 @@ limitations under the License.
+
+	// keyFunc is used to make the key used for queued item insertion and retrieval, and
+	// should be deterministic.
+	keyFunc KeyFunc
```
looks like this is owned by `Heap`. `heapData` isn't using it.

bsalamat replied:
That's a valid point; `headData` does not use `keyFunc`. This is in `heapData` because `heapData` holds all the data that `Heap` needs.

anfernee reviewed (commented):
looks good. some minor suggestions.

bsalamat commented:
@timothysc @davidopp It would be great if you guys could take another look.

davidopp commented on plugin/pkg/scheduler/core/scheduling_queue.go:
```diff
@@ -15,47 +15,679 @@ limitations under the License.
+}
+
+// Add adds a pod to the active queue. It should be called only when a new pod
+// is added.
```
maybe add "(so there is no chance the pod is already in either queue)" - it took me a while to figure out why you had the "It should be called only when..."

davidopp commented on plugin/pkg/scheduler/core/scheduling_queue.go:
```diff
@@ -15,47 +15,679 @@ limitations under the License.
+}
+
+// AddUnschedulable adds a pod to the unschedulable queue.
+func (p *PriorityQueue) AddUnschedulable(pod *v1.Pod) error {
```
I don't claim to understand the code well enough to know, but thought I'd ask anyway: If AddUnschedulable() is called when p.receivedMoveRequest() is true and the pod is currently in the unschedulableQ, should you delete it from unschedulable and add it to activeQ? Or is this always just a bug?

bsalamat replied:
If the pod is already in the unschedulableQ, a move request should have moved it to activeQ.

davidopp reviewed (commented):
I'm still reviewing but was confused about receivedMoveRequest so sending my comments so far now.

davidopp commented on plugin/pkg/scheduler/core/scheduling_queue.go:
```diff
@@ -15,47 +15,679 @@ limitations under the License.
+	return err
+}
+
+// Delete deletes the item from either of the two queues.
```
Maybe add "Assumes the pod is only in one queue."

davidopp commented on plugin/pkg/scheduler/core/scheduling_queue.go:
```diff
@@ -15,47 +15,679 @@ limitations under the License.
+	p.cond.Broadcast()
+}
+
+// movePodsWithMatchingAffinityToActiveQueue moves unschedulable pods which have
```
this comment doesn't describe this function (move vs. get)

davidopp commented on plugin/pkg/scheduler/core/scheduling_queue.go:
```diff
@@ -15,47 +15,679 @@ limitations under the License.
+	return nil
+}
+
+// AssignedPodAdded is called when a bound pod is added. Creation of this pod
```
Is this trying to say "AssignedPodAdded is called when a pod is bound to a node. Scheduling the pod may make pending pods with matching pod affinity terms schedulable"

If so, that would be clearer.

bsalamat replied:
What you said is a more probable scenario, but this event is also triggered when a pod whose node name is set is created.

davidopp commented on plugin/pkg/scheduler/core/scheduling_queue.go:
```diff
@@ -15,47 +15,679 @@ limitations under the License.
+	p.lock.RLock()
+	defer p.lock.RUnlock()
+	podsToMove := []*v1.Pod{}
+	for _, p := range p.unschedulableQ.pods {
```
please don't use "p" as the variable name here, as it's shadowing the other "p"

bsalamat commented on plugin/pkg/scheduler/core/scheduling_queue.go:
```diff
@@ -15,47 +15,679 @@ limitations under the License.
+	unschedulableQ *UnschedulablePodsMap
+	// receivedMoveRequest is set to true whenever we receive a move request, and
+	// is set to false, when we pop a pod. It indicates if we received a move
+	// request when a pod was in flight. In such case, we put the pod back into the
```
The answer to the first question is yes. When scheduler tries to schedule pod A (when pod A is in flight), if we receive a move request to move pods to activeQ, we should probably put pod A to the activeQ if it is determined unschedulable. That's why this flag exists. The flag tracks any moves from unschedulableQ to the activeQ while scheduler attempts to schedule a pod. That's why popping a pod resets the flag to false. It marks the beginning of another cycle.
That said, you are right that the flag is not precise and there is no guarantee that we always see the next Pop() after pod A is put back into the queue.

davidopp replied:
OK everything made sense except the last sentence. What do you mean by "we always see"?

bsalamat replied:
Scheduler updates the status of an unschedulable pod at the end of each cycle, but since receiving the event for the update happens asynchronously (via event handlers), scheduler may Pop() another pod and then the event for adding the pod which was picked in the previous cycle arrives. That's what I meant by the last sentence.

bsalamat commented on plugin/pkg/scheduler/factory/factory.go:
```diff
@@ -331,6 +331,7 @@ func (c *configFactory) onPvAdd(obj interface{}) {
 		}
 		c.invalidatePredicatesForPv(pv)
 	}
+	c.podQueue.MoveAllToActiveQueue()
```
You are right. I will remove this.

bsalamat commented:
@davidopp Addressed the comments. PTAL. Thanks!

davidopp commented on plugin/pkg/scheduler/core/scheduling_queue.go:
```diff
@@ -15,47 +15,694 @@ limitations under the License.
+	}
+}
+
+// Exists returns true if the given pod is found in the map.
```
s/true/the pod/

bsalamat commented on plugin/pkg/scheduler/core/scheduling_queue.go:
```diff
@@ -15,47 +15,694 @@ limitations under the License.
+// is called unschedulableQ.
+// Heap is already thread safe, but we need to acquire another lock here to ensure
+// atomicity of operations on the two data structures..
+type PriorityQueue struct {
```
Yes. The main queue that scheduler uses to pick up pods is a priority queue.

bsalamat commented on plugin/pkg/scheduler/core/scheduling_queue.go:
```diff
@@ -15,47 +15,694 @@ limitations under the License.
+	// pods is a map key by a pod's full-name and the value is a pointer to the pod.
+	pods map[string]*v1.Pod
+	// nominatedPods is a map keyed by a node name and the value is a list of
+	// pods' full-names which are nominated to run on the node.
```
When a pod is nominated, it is marked unschedulable. This is scheduler's default behavior to mark such pods as unschedulable since they cannot be scheduled right away. These pods are added to unschedulableQ and may be moved to the activeQ from time to time when we receive a move request. Scheduler tries to schedule them as soon as possible and does not wait to put them only on the same node as their nominated node.
In my second PR, there is a logic in checking whether a node can fit a pod that picks up the nominated pods from the queue and makes sure that the being tried pods fits in presence of these nominated pods.

bsalamat commented:
Thanks, @davidopp! Addressed the comments. The last fixup commit contains the new changes.

davidopp commented:
I vaguely remember that Borg periodically unconditionally moves things from (the Borg equivalent of) unschedulable queue to active queue, just in case there is a bug that would otherwise leave something in unschedulable forever. It runs off of a timer with a configurable interval (I don't remember the exact interval, but it was something on the order of an hour). Not necessary for this PR but might be good for a followup PR.

bsalamat commented:
@davidopp Yes, I was also thinking about having a similar mechanism. I think I will add it, but as you said, it will be in a follow-up PR.

bsalamat commented:
/retest

davidopp commented:
/lgtm
/approve

bsalamat committed:
Autogenerated files

bsalamat commented:
@davidopp I removed a file outside of the scheduler's code which was reformatted accidentally by my IDE and also squashed commits. Could you please approve again?

davidopp commented:
/lgtm
/approve

I suspect that creating or deleting a PVC cannot make an unschedulable pod schedulable, but we can fix that in a followup PR if that's the case, since this PR is doing the conservative thing.

bsalamat commented:
/retest
